clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -799,64 +804,95 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
+    }
+
+    private Optional<Exception> revokePartitions(SortedSet<TopicPartition> partitions, int generation, String memberId) {
+
         // the generation / member-id can possibly be reset by the heartbeat thread
         // upon getting errors or heartbeat timeouts; in this case whatever is previously
         // owned partitions would be lost, we should trigger the callback and cleanup the assignment;
         // otherwise we can proceed normally and revoke the partitions depending on the protocol,
         // and in that case we should only change the assignment AFTER the revoke callback is triggered
         // so that users can still access the previously owned partitions to commit offsets etc.
-        Exception exception = null;
-        final SortedSet<TopicPartition> revokedPartitions = new TreeSet<>(COMPARATOR);
+        Optional<Exception> exception = Optional.empty();
         if (generation == Generation.NO_GENERATION.generationId ||
-            memberId.equals(Generation.NO_GENERATION.memberId)) {
-            revokedPartitions.addAll(subscriptions.assignedPartitions());
-
-            if (!revokedPartitions.isEmpty()) {
+                memberId.equals(Generation.NO_GENERATION.memberId)) {
+            if (!partitions.isEmpty()) {
                 log.info("Giving away all assigned partitions as lost since generation/memberID has been reset," +
-                    "indicating that consumer is in old state or no longer part of the group");
-                exception = invokePartitionsLost(revokedPartitions);
-
+                        "indicating that consumer is in old state or no longer part of the group");
+                exception = Optional.ofNullable(invokePartitionsLost(partitions));
                 subscriptions.assignFromSubscribed(Collections.emptySet());
             }
         } else {
             switch (protocol) {
                 case EAGER:
-                    // revoke all partitions
-                    revokedPartitions.addAll(subscriptions.assignedPartitions());
-                    exception = invokePartitionsRevoked(revokedPartitions);
-
+                    exception = Optional.ofNullable(invokePartitionsRevoked(partitions));
                     subscriptions.assignFromSubscribed(Collections.emptySet());
 
                     break;
 
                 case COOPERATIVE:
-                    // only revoke those partitions that are not in the subscription any more.
                     Set<TopicPartition> ownedPartitions = new HashSet<>(subscriptions.assignedPartitions());
-                    revokedPartitions.addAll(ownedPartitions.stream()
-                        .filter(tp -> !subscriptions.subscription().contains(tp.topic()))
-                        .collect(Collectors.toSet()));
-
-                    if (!revokedPartitions.isEmpty()) {
-                        exception = invokePartitionsRevoked(revokedPartitions);
+                    partitions.addAll(ownedPartitions.stream()
```

Comment:
Perhaps we could put this in a helper called `revokeUnsubscribedPartitions` and add a comment:
> For the cooperative strategy, partitions are usually revoked in `onJoinComplete` when the coordinator explicitly requests revocation in order to assign them to another group member. However, any partitions of topics which are no long subscribed can be revoked immediately before rejoining the group.

If we do that, perhaps there's no need to mutate `partitions`. We're already building a new collection here, so we can just make it a new SortedSet.